src/solvers/refinement/string_refinement_util.h

```diff
@@ -0,0 +1,315 @@
+  void add(const std::size_t i, const exprt &expr);
+
+  /// \param i: index corresponding to an equation
+  /// \return vector of expressions contained in the equation with the given
+  ///   index `i`
+  std::vector<exprt> find_expressions(const std::size_t i);
+
+  /// \param expr: an expression
+  /// \return vector of equations containing the given expression `expr`
+  std::vector<std::size_t> find_equations(const exprt &expr);
+
+private:
+  /// Record index of the equations that contain a given expression
+  std::map<exprt, std::vector<std::size_t>> equations_containing;
+  /// Record expressions that are contained in the equation with the given index
+  std::unordered_map<std::size_t, std::vector<exprt>> strings_in_equation;
+};
+
+/// Base class for string primitives
+class string_primitivet
+{
+public:
+  string_primitivet() = default;
+  string_primitivet(const string_primitivet &) = delete;
+
+  virtual optionalt<array_string_exprt> string_result() const
+  {
+    return {};
+  }
+
+  virtual std::vector<array_string_exprt> string_arguments() const
+  {
+    return {};
+  }
+};
+
+/// String primitive transforming one string into another
+class string_transformation_primitivet : public string_primitivet
+{
+public:
+  array_string_exprt result;
+  array_string_exprt input;
+  std::vector<exprt> args;
+  exprt return_code;
+  optionalt<array_string_exprt> string_result() const override
+  {
+    return result;
+  }
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return {input};
+  }
+};
+
+/// String inserting a string into another one
+class string_insertion_primitivet : public string_primitivet
+{
+public:
+  array_string_exprt result;
+  array_string_exprt input1;
+  array_string_exprt input2;
+  std::vector<exprt> args;
+  exprt return_code;
+
+  /// Constructor from arguments of a function application
+  string_insertion_primitivet(
+    const std::vector<exprt> &fun_args,
+    array_poolt &array_pool);
+
+  optionalt<array_string_exprt> string_result() const override
+  {
+    return result;
+  }
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return {input1, input2};
+  }
+};
+
+/// String creation from other types
+class string_creation_primitivet : public string_primitivet
+{
+public:
+  array_string_exprt result;
+  std::vector<exprt> args;
+  exprt return_code;
+
+  optionalt<array_string_exprt> string_result() const override
+  {
+    return result;
+  }
+};
+
+/// String test
+class string_test_primitivet : public string_primitivet
+{
+public:
+  exprt result;
+  std::vector<array_string_exprt> under_test;
+  std::vector<exprt> args;
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return under_test;
+  }
+};
+
+/// Keep track of dependencies between strings.
+/// Each string points to primitive calls on which it depends,
+/// each primitive points to the strings on which the result depend.
+class dependency_grapht
```

Comment:
Is there a really good reason not to use/inherit from `grapht`?

Reply:
We have two different types of nodes (`primitive_nodet` and `string_nodet`), and we do not need to store information on edges, so using `grapht` seems like more work.

Reply:
I think you should reconsider; grapht will provide you with useful graph algorithms "off the peg" which you will surely want in time. Either encode your two node types as a class hierarchy (so you can use dynamic_cast to determine the type of a given node) or as a more- or less-formal union type.

Reply:
Sounds like a slightly weird sort of graph (different nodes, no edges) :-)  I just want to make sure we keep the amount of duplicate code to a minimum. Not-invented-here syndrome is a pretty dangerous disease.

Reply:
@smowton 
> I think you should reconsider; grapht will provide you with useful graph algorithms "off the peg" which you will surely want in time. Either encode your two node types as a class hierarchy (so you can use dynamic_cast to determine the type of a given node) or as a more- or less-formal union type.

Not sure I will need these algorithms, since this is an acyclic graph, computing SCCs is useless, so is `is_dag`, and `shortest_loop`. `shortest_path` is not really meaningful in this context. Maybe `get_reachable` would make sense but anyway I don't like to have to inherit from a class to be able to use its algorithms. I'd prefer the algorithms to be implemented as function templates, here for instance for `get_reachable`: https://github.com/romainbrenguier/cbmc/pull/5

Reply:
Please do bring in changes like https://github.com/romainbrenguier/cbmc/pull/5 - otherwise "you're not doing it my way so I don't use it" sounds like not-invented-here syndrome has already manifested. The code base is pretty big already, and someone will have to continue to maintain all these copies of almost-the-same-but-not-quite-the-same code.

Reply:
I might be misunderstanding, so please correct me if I'm wrong: it seems you suggest to retain your own data structure, but intend to use functions that otherwise are only used on `grapht` (moving them from member functions to free functions)?

If the above is right, the best possible picture is making all functions that work on those graph-like objects apply to anything that shares the same API as `grapht`.

Reply:
@tautschnig 
> I might be misunderstanding, so please correct me if I'm wrong: it seems you suggest to retain your own data structure, but intend to use functions that otherwise are only used on grapht (moving them from member functions to free functions)?

That's right

> If the above is right, the best possible picture is making all functions that work on those graph-like objects apply to anything that shares the same API as grapht.

The public interface of `grapht` is huge so I don't wan't to re-implement all these, just so that maybe one day I can reuse one basic function like `get_reachable`.

Reply:
If the public interface of `grapht` is so large as to hinder its reuse, is there any milage in splitting up it's interface into a finer grained class hierarchy?

Reply:
All computer applications consist of graphs. But each has different needs; some are trees, some are lists, some are directed (like this one), some are not (like `grapht`). Directed graphs can be easily expressed with `shared_ptr`s.

`grapht` has too many unnecessary features for it to be useful here, and we shouldn't modify it just to suit this one particular use case.

Reply:
This looks quite specialised; still, it could be based on `grapht` in the way @smowton suggests. 
In particular, I don't understand that argument:

> The public interface of grapht is huge so I don't wan't to re-implement all these

Why would you need to reimplement any of these?

Reply:
How is the graph going to be used? At the moment, it seems you can only print it.
Maybe this will shed some light on which algorithms from `grapht` you'll reuse/reimplement.

Reply:
@chrisr-diffblue you would never reimplement its public interface; the typical usage looks like this: https://github.com/diffblue/cbmc/blob/develop/src/analyses/call_graph.h#L113

Basically you define an edge struct (or don't if, like this graph, edges are unlabelled), then a node struct (here a tagged union of intrinsic function calls and strings), then finally inherit grapht<my_node_typet>. It gives you standard mutators and accessors for making nodes, adding / querying / deleting edges, and so on, plus a small library of graph algorithms that work over a grapht.

For this use case you'd do something like
```
struct string_dependence_graph_nodet : graph_nodet {
  exprt string_or_intrinsic; // Use .id() to differentiate them if context isn't enough
};
struct string_dependence_grapht : grapht<string_dependence_graph_nodet>
{
} mygraph;

auto new_node_idx = mygraph.add_node();
mygraph[new_node_idx].string_or_intrinsic = some_expr;
...
```

The advantage is that that you get things like `mygraph.get_reachable(some_node_idx, forwards)` for free, and so avoid writing yet another common graph algorithm implementation for each type that needs it.

I agree that this sort of thing should be done as free functions expecting a concept, a la STL algorithms, but it doesn't make a massive difference to how you use them -- you'd still need to conform to an interface somehow by implementing `get_node`, `get_edges` and similar, probably by inheriting some useful mixin.

Whether this use case will ever **want** the standard graph algorithms though? @romainbrenguier thinks no, I think we might as well describe this as a generic graph just in case we want them in the future since it's little effort to do so, but I don't have a strong enough opinion on the matter to block on this.

Reply:
@peterschrammel 
> In particular, I don't understand that argument:
> > The public interface of grapht is huge so I don't wan't to re-implement all these

> Why would you need to reimplement any of these?

This wasn't an argument against inheriting from grapht, but against "making all functions that work on those graph-like objects apply to anything that shares the same API as grapht".

Reply:
@peterschrammel 
> Now, we have a dependence_grapht and a dependency_grapht in the code base. I don't think that this a good idea.

What about `string_dependence_grapht` or just `string_dependencest`?

Reply:
@romainbrenguier, either is fine.

How is the graph going to be used?

Reply:
We will look at the set of primitives that are "relevant" for the solver: that is all test and comparison primitives and their dependencies. These will get added as string_constraints but not the others (the remaining primitives do not affect satisfiability). The algorithm we need for that would be some kind of `get_reachable` which takes a set of nodes as inputs instead of a single node. I can write that as a function template in `util/graph` so that it can be reused by any graph-like structure.